### PR TITLE
match stale solutions to correct header

### PR DESCRIPTION
**Pull request by nerdralph**
Partial fix for 567.
https://github.com/ethereum-mining/ethminer/issues/567
Stales aren't flagged as stale yet, but the nonce is matched to the proper WorkPackage.

### CODE REVIEW

**Comment by jean-m-cyr** [comment]
For what it's worth, I merged this with the nolock PR, just for grins, and ran it in cl mode for 50 shares.
[A50+0:R0+0:F0] Time: 00:43 (4x1060, Ubuntu)
All looks very normal. No stales of course, but also no rejects. No experience with cl though and how often these mismatches might occur. I was more interested in seeing that the nolock patch doesn't interfere with the cl miner algorithm.

**Comment by nerdralph** [comment]
Well under 1% of shares would be stale.  You can calculate the probability
from the hashrate, worksize, and difficulty.

File: libethash-cl/CLMiner.cpp [hunk 1]
```diff
@@ -384,6 +380,11 @@ void CLMiner::workLoop()
 			if (nonce != 0)
 				report(nonce, current);
 
+			current = w;        // kernel now processing newest work
+            current.startNonce = startNonce;
```
Comment (chfast) [inline libethash-cl/CLMiner.cpp]: 1. The indentation is off.
2. If we are going to keep the "start nonce" in `current`, maybe the free `startNonce` is not needed.

File: libethash-cl/CLMiner.cpp [hunk 2]
```diff
@@ -384,6 +380,11 @@ void CLMiner::workLoop()
...
+			// Increase start nonce for following kernel execution.
```
Comment (chfast) [inline libethash-cl/CLMiner.cpp]: Update comment: "following" -> "next".

**Comment by nerdralph** [comment]
I'll fix the tab/space.
My first attempt was without the startNonce variable, but it didn't work
because w gets overwritten at the top of the while loop.  It would take
more comex changes to the code in order get rid of it.

**Comment by nerdralph** [comment]
I meant more complex changes.

File: libethash-cl/CLMiner.cpp [hunk 3]
```diff
@@ -731,14 +730,19 @@ bool CLMiner::init(const h256& seed)
...
 			m_dagKernel.setArg(0, i * m_globalWorkSize);
 			m_queue.enqueueNDRangeKernel(m_dagKernel, cl::NullRange, m_globalWorkSize, m_workgroupSize);
 			m_queue.finish();
-			cllog << "DAG" << int(100.0f * i / fullRuns) << '%';
 		}
+		auto endDAG = std::chrono::steady_clock::now();
 
+		auto dagTime = std::chrono::duration_cast<std::chrono::milliseconds>(endDAG-startDAG);
+		char gb[16];
+		sprintf(gb, "%.2f", (float)dagSize / (1024 * 1024 * 1024));
```
Comment (chfast) [inline libethash-cl/CLMiner.cpp]: WTF? Just do `cnote << std::setprecision(2) << f << " GB of DAG..."`

Comment (nerdralph) [reply libethash-cl/CLMiner.cpp]: Not sure why this is showing under this pull as I did it as a separate pull request.
As for why, some of us learned C++ before std::setprecision() existed; I did most of my C++ development 20-30 years ago, having gotten started with cfront.
Anyway, I'll update the code to use setprecision shortly.

Comment (nerdralph) [reply libethash-cl/CLMiner.cpp]: And it seems cnote doesn't work with the precision iomanip:
"2.27344  GB of DAG data generated in 3886 ms"
I'm not about to fix cnote, so I'll just leave it as the default float output.

**Pull request merged**